SNOW-1183322: [Local Testing] Add support for registering sprocs

sfc-gh-jrose opened the pull request:
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #1183322

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

sfc-gh-aling commented on src/snowflake/snowpark/mock/_stored_procedure.py:
```diff
@@ -0,0 +1,174 @@
+            else:
+                parsed_args.append(arg)
+
+        # import pdb; pdb.set_trace()
```
nit: remove?
```suggestion
        # import pdb; pdb.set_trace()
```

sfc-gh-aling commented on src/snowflake/snowpark/mock/_stored_procedure.py:
```diff
@@ -0,0 +1,174 @@
+                        f"Unexpected type {expr.datatype} for sproc argument of type {expected_type}"
+                    )
+                parsed_args.append(
+                    calculate_expression(
+                        expr,
+                        ColumnEmulator(data=[None]),
+                        session._analyzer,
+                        {},
+                        keep_literal=True,
```
I have two dumb questions here:

1. what is `Column` type args for sproc and how this is used by users?
2. why we set `ColumnEmulator(data=[None]),` here, this gives us an empty column with just the name, is this intended? this is kinda relavant to the first question on how `Column` is used in sproc.

sfc-gh-jrose replied:
1. Some of our tests pass in `lit` values rather than raw values. Those are the only Column objects I've observed being passed to sproc.
2. The way calculate_expression works for lit values requires there to be some input data even if it is all null. I think I can refactor this so that it is more clear.

sfc-gh-stan replied:
These are good questions actually. In fact I feel like we shouldn't support column inputs to stored procedures, although that might be a limitation from `DataFrame.select`...

sfc-gh-stan replied:
+1 for refactor to be more clear

sfc-gh-aling commented on src/snowflake/snowpark/mock/_stored_procedure.py:
```diff
@@ -0,0 +1,174 @@
+        source_code_display: bool = True,
+        skip_upload_on_content_match: bool = False,
+    ) -> StoredProcedure:
+        raise NotImplementedError(
+            "[Local Testing] Registering stored procedure from file is not currently supported."
+        )
```
let's use the telemetry not supported api.
also is this planned be supported by @sfc-gh-stan soon?

sfc-gh-stan replied:
Yea I'll have a separate PR that adds the support for imports.

sfc-gh-aling commented on src/snowflake/snowpark/mock/_stored_procedure.py:
```diff
@@ -0,0 +1,174 @@
+        check_python_runtime_version(self._session._runtime_version_from_requirement)
+
+        if udf_name in self._registry:
+            raise SnowparkSQLException(
```
with parameter `replace`, user can overwrite the existing one

sfc-gh-aling commented on src/snowflake/snowpark/mock/_stored_procedure.py:
```diff
@@ -0,0 +1,174 @@
+            anonymous,
+        )
+
+        if is_pandas_udf:
+            raise TypeError("pandas stored procedure is not supported")
+
+        if packages or imports:
+            raise NotImplementedError(
+                "[Local Testing] Uploading imports and packages not currently supported."
+            )
```
let's use oob telemetry/mock connection log not supported api

sfc-gh-stan commented on src/snowflake/snowpark/mock/_stored_procedure.py:
```diff
@@ -0,0 +1,176 @@
+        # Unpack columns if passed
+        parsed_args = []
+        for arg, expected_type in zip(args, self._input_types):
+            if isinstance(arg, Column):
```
Question: why do we need this case? Do we support column as a stored procedure argument?

sfc-gh-stan commented on tests/integ/test_stored_procedure.py:
```diff
@@ -161,54 +163,100 @@ def return1(session_):
     assert int2str_sp(123, session=session) == "123"
-    assert pow_sp(2, 10, session=session) == 1024
 
+    if local_testing_mode is False:
```
```suggestion
    if not local_testing_mode:
```

sfc-gh-stan commented:
Can we enable more tests in `tests/integ/test_stored_procedure.py` like `test_type_hints`?

sfc-gh-jrose commented:
I've enabled all the tests that look relevant to local testing. There's several more tests related to imports and loading from file that can't be enabled yet. Turning on those tests revealed a couple of gaps. 

1. session.call didn't work previously, but should now. 
2. Calling a sproc with a nested column expression that returns a single value appears to be supported in live so I've reverted the refactor I made to add support for calling `calculate_expression` back. I've added a comment so that it's a little more understandable.

In order to implement tests 1:1 I also implemented functions.sqrt and functions.pow. They are trivial so shouldn't expand the scope of this PR by much.

sfc-gh-aling commented on CHANGELOG.md:
```diff
@@ -25,9 +25,12 @@
     - object_construct
     - object_construct_keep_null
+    - pow
+    - udf
+    - sproc
```
nit: I think udf and sproc are worth being mentioned separately in the changelog

sfc-gh-aling commented on src/snowflake/snowpark/mock/_stored_procedure.py:
```diff
@@ -0,0 +1,238 @@
+        # Unpack columns if passed
+        parsed_args = []
+        for arg, expected_type in zip(args, self._input_types):
+            if isinstance(arg, Column):
+                expr = arg._expression
+
+                # If expression does not define its datatype we cannot verify it's compatibale.
+                # This is potentially unsafe.
+                if expr.datatype and not sproc_types_are_compatible(
+                    expr.datatype, expected_type
+                ):
+                    raise ValueError(
+                        f"Unexpected type {expr.datatype} for sproc argument of type {expected_type}"
+                    )
+
+                # Expression may be a nested expression. Expression should not need any input data
+                # and should only return one value so that it can be passed as a literal value.
+                # We pass in a single None value so that the expression evaluator has some data to
+                # pass to the expressions.
+                resolved_expr = calculate_expression(
+                    expr,
+                    ColumnEmulator(data=[None]),
+                    session._analyzer,
+                    {},
+                )
+
+                # If the length of the resolved expression is not a single value we cannot pass it as a literal.
+                if len(resolved_expr) != 1:
+                    raise ValueError(
+                        "[Local Testing] Unexpected argument type {expr.__class__.__name__} for call to sproc"
+                    )
+                parsed_args.append(resolved_expr[0])
```
can you give me an example of stored procedure with one `arg` being Column? I'm trying to understand how this arg is used -- is it used to getting column of data within this process or we just need to parse the expression into a literal.

if this is just analyzing an expression to infer a literal value, then the MockAnalyzer.analyze can better help here: https://github.com/snowflakedb/snowpark-python/blob/main/src/snowflake/snowpark/mock/_analyzer.py#L169-L176.

sfc-gh-jrose replied:
This test has an example of where this is used:
https://github.com/snowflakedb/snowpark-python/pull/1338/files#diff-4963488bfdaad538e6af7222d4db977758e66d05b4c11b600583449152416f50L221

sfc-gh-aling replied:
I see, if evaluating the literal value is the only case, then let's try using `session._analyzer.analyze`.

calculate_expression is for calculating table/column results, while analyze is just for expressions parsing

sfc-gh-jrose replied:
Running analyze on this statement produces query text which is not what the function expects as input. For example, the sqrt done in the test when put through the analyzer creates the string `sqrt(36)` which is not what the function expects as input.

sfc-gh-aling replied:
I see, that makes sense. thanks for checking it!

sfc-gh-stan commented on src/snowflake/snowpark/mock/_stored_procedure.py:
```diff
@@ -54,20 +67,51 @@ def __call__(
+
+                # If expression does not define its datatype we cannot verify it's compatibale.
+                # This is potentially unsafe.
+                if expr.datatype and not sproc_types_are_compatible(
```
Question: in what case is expr.datatype empty here?

sfc-gh-jrose replied:
The datatype of the sqrt function in this example is None:
https://github.com/snowflakedb/snowpark-python/pull/1338/files#diff-4963488bfdaad538e6af7222d4db977758e66d05b4c11b600583449152416f50L221

sfc-gh-stan commented on src/snowflake/snowpark/mock/_stored_procedure.py:
```diff
@@ -31,11 +42,13 @@
-    elif isinstance(x, _IntegralType) and isinstance(y, _IntegralType):
-        return True
-    elif isinstance(x, _FractionalType) and isinstance(y, _FractionalType):
+    if (
```
When we work on coercion, this function should be replaced by like `DataType.can_coerce_to(DataType)`. @sfc-gh-aling

sfc-gh-stan commented on src/snowflake/snowpark/mock/_stored_procedure.py:
```diff
@@ -0,0 +1,238 @@
+                MapType,
+                StructType,
+            ),
+        ) and not isinstance(result, DataFrame):
```
We have no support for table sprocs yet, why do we need `not isinstance(result, DataFrame)`?

sfc-gh-jrose replied:
I've seen examples in the tests where a sproc returns a Dataframe, but it's signature says it returns a Structtype. If it's truly a struct type it needs to be serialized, but if it's just a dataframe it does not.

sfc-gh-stan replied:
I see. Could you please point me to the test?

sfc-gh-jrose replied:
See [here](https://github.com/snowflakedb/snowpark-python/blob/e7421b44623e4388fe4f31596ca65bd45abf8109/tests/integ/test_stored_procedure.py).

sfc-gh-stan commented:
LGTM overall, had a couple of questions and comments.

sfc-gh-aling commented:
I noticed the oob telemetry test is flaky these days thus blocking the merge gate.
I created a PR to fix the flakiness: https://github.com/snowflakedb/snowpark-python/pull/1382
